lib/dry/container/mixin.rb

```diff
@@ -235,6 +239,18 @@ def import(namespace)
       def _container
         @_container
       end
+
+      def _registry
+        respond_to?(:registry) ? registry : self.class.registry
+      end
+
+      def _resolver
+        respond_to?(:resolver) ? resolver : self.class.resolver
+      end
+
+      def _namespace_separator
+        respond_to?(:namespace_separator) ? namespace_separator : self.class.namespace_separator
+      end
```

Comment:
@GustavoCaso What's the thinking behind these `respond_to?` checks, rather than just delegating directly to `self.class`?

Reply:
I check using `respond_to?` because since `dry-container` could be used at class level and instance level, when the caller is at class level we get this error `undefined method `resolver' for Class:Class`

I could check instead of using `respond_to?` I could check whether is a class or an instance.

WDYT? @timriley @solnic

Reply:
@GustavoCaso Ah yeah, I see the issue now.

My first thought was that maybe we should just "bless" a single type of usage for the dry-container mixin – either include or extend – but I think there are valid use cases for both :)

So, if instead we just want to find a way for these internal methods to read a little nicer, how about something like this?

```ruby
def namespace_separator
  defined?(super) ? super : self.class.namespace_separator
end
```

By using the `defined?(super)` check and calling super if it's there, it means at least we no longer need to prefix these method names with an underscore.

Another approach could be to invert the check:

```ruby
def namespace_separator
  self.class.respond_to?(:namespace_separator) ? self.class.namespace_separator : super
end
```

Whichever approach we take, I think it might be good to prefix all these similarly-behaving methods with a comment explaining why they work like they do.

Reply:
We should make it in a way that there's no need for any checks at run-time. Which means a container class shouldn't have these methods at all, as they are provided by `ClassAttributes`, and a container object should have them implemented and they would simply delegate to `self.class`.

Reply:
@timriley Thank you so much for the answer.

I never use or thought about the possibility of `defined?(super)` but after your explanation makes a lot of sense. For clarity reason I would stick with the second approach I think it reads better for these case.

Also totally agree with the comment to also help understand the reason behind them.

Reply:
@GustavoCaso Nice outcome in https://github.com/dry-rb/dry-container/pull/41/commits/ee0877a0e09d927f72f4b45f48b161fca3b73e92 :)